src/main/java/org/broadinstitute/consent/http/service/UserService.java

```diff
@@ -454,12 +454,127 @@ public void hasValidActiveERACredentials(User user) {
+        && hasAssignedInstitutionInDatabase
+        && !hasMatchingInstitutionInDatabase) {
+      return updateInstitutionForUserAndReturnUpdatedUser(user, institutionFromEmail.getId());
+    }
+    // case 8
+    if (hasInstitutionMatchingEmailDomain
+        && !hasLibraryCard
+        && !hasAssignedInstitutionInDatabase
+        && !hasMatchingInstitutionInDatabase) {
+      return updateInstitutionForUserAndReturnUpdatedUser(user, null);
+    }
+    // case 10
+    if (!hasInstitutionMatchingEmailDomain
+        && hasLibraryCard
+        && hasAssignedInstitutionInDatabase
+        && !hasMatchingInstitutionInDatabase) {
+      return dropLCAndInstitutionForUser(user);
+    }
+    // case 12
+    if (!hasInstitutionMatchingEmailDomain
+        && hasLibraryCard
+        && !hasAssignedInstitutionInDatabase
+        && !hasMatchingInstitutionInDatabase) {
+      dropLibraryCardForUser(user);
+      return findUserByEmail(user.getEmail());
+    }
+    // case 14
+    if (!hasInstitutionMatchingEmailDomain
+        && !hasLibraryCard
+        && hasAssignedInstitutionInDatabase
+        && !hasMatchingInstitutionInDatabase) {
+      return updateInstitutionForUserAndReturnUpdatedUser(user, null);
+    }
+
+    // do nothing cases 1, 5, 15, 16
+    return user;
+
+    // not possible cases: 3, 7, 9, 11, 13
+  }
+
+  private void updateInstitutionForUser(User user, Integer institutionId) {
+    userDAO.updateInstitutionId(user.getUserId(), institutionId);
+  }
+
+  private User dropLCAndInstitutionForUser(User user) {
+    updateInstitutionForUser(user, null);
+    libraryCardDAO.deleteAllLibraryCardsByUser(user.getUserId());
+    return findUserByEmail(user.getEmail());
+  }
+
+  private User updateInstitutionForUserAndReturnUpdatedUser(User user, Integer institutionId) {
+    updateInstitutionForUser(user, institutionId);
+    return findUserByEmail(user.getEmail());
+  }
+
+  private User dropLibraryCardForUser(User user) {
+    libraryCardDAO.deleteAllLibraryCardsByUser(user.getUserId());
+    return findUserByEmail(user.getEmail());
+  }
+
+  @VisibleForTesting
+  protected boolean hasInstitutionMatchingEmailDomain(Institution institution) {
+    return institution != null;
+  }
+
+  @VisibleForTesting
+  protected boolean hasLibraryCard(User user) {
+    return !user.getLibraryCards().isEmpty();
+  }
+
+  @VisibleForTesting
+  protected boolean hasMatchingInstitutionInDatabase(
+      Institution institutionFromEmail, Institution institutionFromDatabase) {
+    if (institutionFromEmail == null || institutionFromDatabase == null) {
+      return false;
+    }
+    return institutionFromDatabase.equals(institutionFromEmail);
+  }
+
   public static class SimplifiedUser {
 
-    public Integer userId;
-    public String displayName;
-    public String email;
-    public Integer institutionId;
+    private Integer userId;
+    private String displayName;
+    private String email;
+    private Integer institutionId;
```

Comment:
Looking back on this class, it really should be a record.

Reply:
I was thinking the same thing when I made the last getter but didn't know if that was too much of a refactor right now.  I may take a stab at it when I handle the NFE.

Reply:
Looks like this will be more involved because of changes we'll need to make with the constructor that maps a user.  Not a high bar, but not worth the squeeze at the moment.

Reply:
If you're not going to convert this to a record now, I wouldn't bother with any cleanups in this PR, as any changes you make will end up being changed again when the record migration is done.

Having a custom constructor for a record is easy to do. Also, if you modify the code for `SimplifiedUser` slightly, you can enable the intellij refactoring operation to convert it to a `record` which takes care of getters etc automatically for you.